File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public long  cal(int[] piles, int speed){
         long hour = 0;
        for(int it: piles){
            if(it % speed == 0){
                hour += it/speed;
            }else {
                 hour += it/speed + 1; 
            }
        }
        return hour;
    }
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1; // koko eating 1 bana per hour 
        int right = 0;// maximum speed to finish in h hour 
        for(int it: piles){
            right = Math.max(it, right);
        }

        int speed = 0;
        while(left <= right){
            int randomSpeed = left + (right - left)/2;
             long hourNeeded = cal(piles,randomSpeed);

            if(hourNeeded <= h){
                speed = randomSpeed;
                right = randomSpeed-1;
            }else{
                left = randomSpeed+1;
            }
        }
        return speed;
    }
}

File: 907-koko-eating-bananas/koko-eating-bananas.cpp
class Solution {
public:
    long long hourWithSpeed(vector<int>& piles, int speed){
        long long hour = 0;
        for(auto it: piles){
            hour += it/speed;
            hour += it%speed == 0 ? 0 : 1;
        }
        return hour;
    }
    int minEatingSpeed(vector<int>& piles, int h) {
        int n = piles.size();
        
        int left = 1, right = piles[0];
        for(auto it: piles){
            right = max(right, it);
        }
        int ans = INT_MAX;
        while(left <= right){
            long long randomSpeed = left + (right - left)/2;
            long long hourNeeded = hourWithSpeed(piles, randomSpeed);

            if(hourNeeded <= h){
                ans = randomSpeed;
                right = randomSpeed-1;
            }else{
                left = randomSpeed+1;
            }
        }
        return ans;
        
    }
};